### DISCUSSION TITLE
Can "polygon.contains(multipoint)" return a vector?

### DISCUSSION THREAD

**Discussion by mkuemmel** [shapely/shapely#1717] [Q&A] [answered]
I need to check for many (millions) of points whether they are in a polygon. Of course this works in shapely, but it should be quicker. The
problem boils down to:
solution = polygon.contains(multipoint)
with solution **not** a boolean (indicating if Polygon contains any of the Points) but a vector indicating which point is in the polygon. Right now I have to go point by point ("polygon.contains(point)"), which is rather slow. Getting all the point specific information in one go might be quick enough.

**Comment by brendan-ward** [comment]
There are a couple ways to do this, but first you have to split the multipoint into an array of points via [get_parts](https://shapely.readthedocs.io/en/stable/reference/shapely.get_parts.html#shapely.get_parts).


From there, you could use STRtree to do both spatial index and containment testing in one go:

```python
points = shapely.get_parts(multipoint)
tree = shapely.STRtree(points)
contained_indexes = tree.query(poly, predicate='contains')

# contained_indexes are integer indexes into the array of points that are contained
```

This uses the index to only test points that overlap with the extent of your polygon, prepares them (for faster predicate), then tests via the `contains` predicate.  All that happens in C, so it is pretty fast.

You could also do this more manually without prefiltering using the spatial index, but this may be slower (untested):

```python
points = shapely.get_parts(multipoint)
shapely.prepare(poly)
contained = shapely.contains(poly, points)

# contained is a boolean index against the array of points
```

**Reply by mkuemmel** [reply]
That's reads quite promising.
I have only v1.8.0, and: "tree.query(poly, predicate='contains')" does not works since the parameter "predicate" does not exist.

Does this also work in v1.8.0? I tried to fund the v1.8 documentation but failed.

**Comment by brendan-ward** [comment]
The vectorized operations here were added in Shapely 2.

The 1.8 docs are at: https://shapely.readthedocs.io/en/1.8.5/

If you go to the [latest docs](https://shapely.readthedocs.io/en/stable/) there is a version dropdown in the bottom left.

**Comment by mkuemmel** [comment]
Okay, so the way I had it was:
`is_in = [my_poly.contains(act_point) for act_point in point_list`

with STRtree in v1.8 this corresponds to:
`point_tree = STRtree(point_list)`
`in_points = [o for o in point_tree.query(my_poly) if my_poly.contains(o)]`

I timed both loops, and actually the version I had (**withoutSTRtree()**) is almost twice as fast (40s vs72s).

Does that make sense? Should V2.0 work quicker than v1.8?

**Comment by brendan-ward** [comment, marked as answer]
Many things are much faster in Shapely 2 with the vectorized methods.

However, in this case - likely depending very much on your data - the contains check without STRtree is about 10x faster (GEOS 3.11):

```python
from timeit import timeit
import shapely
import numpy as np

num_pts = 1000000
points = shapely.points(np.random.random(num_pts), np.random.random(num_pts))
poly = shapely.buffer(shapely.Point(0, 0), 0.5)
shapely.prepare(poly)

# 2.55s
timeit('ix = shapely.STRtree(points).query(poly, predicate="contains")', globals=globals(), number=10)

# 0.28s
timeit('ix = shapely.contains(poly, points)', globals=globals(), number=10)
```

Apologies for not testing that first; I'm just so used to using STRtree as a first pass for many operations like this (though arguably with more complex geometries on either side of comparison).

This is a simple contrived polygon for testing, so there is no benefit to prefiltering by STRtree and it just adds overhead.  However, if your polygon was much more complex (where `contains` predicate grows more complex / slower) and more points fall clearly outside the bounding box of the polygon, there may be a threshold at which prefiltering provides better performance by reducing the number of point in polygon tests.

It is also possible that prefiltering with the bounding box of your polygon will also help, .e.g.

```python
box = shapely.box(*poly.bounds)
ix1 = shapely.contains(box, points)
ix2 = shapely.contains(poly, points[ix])
```

This is 2x slower than plain contains above though.

**Reply by mkuemmel** [reply]
The polygon have only four corners, hence very simple.

Thanks a lot for all the help! My project is going into a hot phase, so I am not sure whether upgrading to 2.0 will be possible. But I'll keep an eye on it and do some testing and timing.

I'll mark this as answered
